### PR TITLE
JAMES-2865 Mock SMTP Server with simple mocking setups

### PR DESCRIPTION
**Pull request by trantienduchn**
Still be drafted but I really need reviews on the structure of code to be able to not doing more mistakes further

### CODE REVIEW

File: server/mailet/mock-smtp-server/pom.xml [hunk 1]
```diff
@@ -60,5 +64,20 @@
...
+            <artifactId>awaitility</artifactId>
+            <scope>test</scope>
+        </dependency>
+        <dependency>
+            <groupId>org.subethamail</groupId>
+            <artifactId>subethasmtp</artifactId>
+            <version>3.1.7</version>
+        </dependency>
+        <dependency>
+            <groupId>org.apache.james</groupId>
```
Comment (Arsnael) [inline server/mailet/mock-smtp-server/pom.xml]: that dependency should be up with the ${james.groupId} groupId

Comment (trantienduchn) [reply server/mailet/mock-smtp-server/pom.xml]: thanks! if there is no more comments, I will fix it and rebase + squash

**Comment by trantienduchn** [comment]
```
[d33518d3a6ea5cc5b6d9d5a42203af587acf45da] [INFO] There is 1 error reported by Checkstyle 8.18 with checkstyle.xml ruleset.
[d33518d3a6ea5cc5b6d9d5a42203af587acf45da] [ERROR] src/main/java/org/apache/james/mock/smtp/server/MockSMTPServer.java:[48,9] (whitespace) WhitespaceAround: WhitespaceAround: 'if' is not followed by whitespace. Empty blocks may only be represented as {} when not part of a multi-block statement (4.1.3)
```

**Comment by trantienduchn** [comment]
rebased to adapt new POJO models

**Comment by chibenwa** [comment]
test this please (restart)

**Comment by trantienduchn** [comment]
rebase master, nothing else

**Comment by chibenwa** [comment]
`EventDeadLettersIntegrationTest.multipleFailedEventsShouldBeCorrectlyProcessedByListenerAfterSuccessfulGroupRedelivery:396 » ConditionTimeout` test this please

**Comment by chibenwa** [comment]
Please rebase this work...

**Comment by Arsnael** [comment]
```
[ERROR]   SpamAssassinContract.expungingSpamMessageShouldNotImpactSpamAssassinState:493 » ConditionTimeout
```

File: server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/MockSMTPServer.java [hunk 2]
```diff
@@ -0,0 +1,69 @@
...
+    Port getPort() {
+        return Port.of(server.getPort());
+    }
+
+    void stop() {
+        server.stop();
+        clearBehavior();
+    }
+
+    void addBehavior(MockSMTPBehavior behavior) {
```
Comment (chibenwa) [inline server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/MockSMTPServer.java]: I'd prefer us to set `MockSMTPBehaviors`

File: server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/MockSMTPServer.java [hunk 3]
```diff
@@ -0,0 +1,69 @@
...
+import org.apache.james.mock.smtp.server.model.Mail;
+import org.apache.james.mock.smtp.server.model.MockSMTPBehavior;
+import org.apache.james.util.Port;
+import org.subethamail.smtp.server.SMTPServer;
+
+class MockSMTPServer {
+
+    private final SMTPServer server;
+    private final ReceivedMailRepository mailRepository;
+    private final List<MockSMTPBehavior> behaviors;
```
Comment (chibenwa) [inline server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/MockSMTPServer.java]: You should rely on the `SMTPBehaviorRepository` for this...

File: server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/MockSMTPServer.java (hunk 2 above)
Comment (chibenwa) [inline server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/MockSMTPServer.java]: addBehavior & clearBehavior should not be a responsability of this component.

File: server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/MockSmtpBehaviors.java [hunk 4]
```diff
@@ -22,6 +22,8 @@
 import java.util.List;
 import java.util.Objects;
 
+import org.apache.james.mock.smtp.server.model.MockSMTPBehavior;
```
Comment (chibenwa) [inline server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/MockSmtpBehaviors.java]: Is that an unused import?

Comment (trantienduchn) [reply server/mailet/mock-smtp-server/src/main/java/org/apache/james/mock/smtp/server/MockSmtpBehaviors.java]: no, that class was located int model package

File: server/mailet/mock-smtp-server/src/test/java/org/apache/james/mock/smtp/server/ReceivedMailRepositoryTest.java [hunk 5]
```diff
@@ -0,0 +1,105 @@
...
+import org.apache.james.mock.smtp.server.model.Mail;
+import org.apache.james.util.MimeMessageUtil;
+import org.junit.jupiter.api.BeforeEach;
+import org.junit.jupiter.api.Test;
+
+import com.google.common.collect.ImmutableList;
+
+class ReceivedMailRepositoryTest {
+    private ReceivedMailRepository testee;
+    private Mail mail;
```
Comment (blackheaven) [inline server/mailet/mock-smtp-server/src/test/java/org/apache/james/mock/smtp/server/ReceivedMailRepositoryTest.java]: can not them be constant?

File: server/mailet/mock-smtp-server/src/test/java/org/apache/james/mock/smtp/server/model/MockSMTPBehaviorTest.java [hunk 6]
```diff
@@ -29,6 +29,9 @@
 import static org.assertj.core.api.Assertions.assertThat;
 import static org.assertj.core.api.Assertions.assertThatThrownBy;
 
+import java.util.Optional;
+
+import org.apache.james.mock.smtp.server.model.Response.SMTPStatusCode;
```
Comment (blackheaven) [inline server/mailet/mock-smtp-server/src/test/java/org/apache/james/mock/smtp/server/model/MockSMTPBehaviorTest.java]: are they used?

File: server/mailet/mock-smtp-server/src/test/java/org/apache/james/mock/smtp/server/ReceivedMailRepositoryTest.java (hunk 5 above)
Comment (chibenwa) [reply server/mailet/mock-smtp-server/src/test/java/org/apache/james/mock/smtp/server/ReceivedMailRepositoryTest.java]: No, MailAddress throws an exception and is really anoying. The alternative is initialization in a static block. I prefer the current solution.

**Comment by chibenwa** [comment]
Merged

**Pull request closed without merging**